MPC: Reduce Memory Allocation

pCYSl5EDgo opened the pull request:
# Background

Current `MessagePack.GeneratorCore` uses Runtime T4 Template.
The limitation of Runtime T4 Template causes problems.

- `MessagePackCompiler.Generator.StringKeyFormatterDeserializeHelper` cannot directly append codes
- `string TransformText()` is stateful
- `File.WriteAllText` is synchronous and not performant

# Methods

- [Cysharp/ZString](https://github.com/cysharp/zstring)
  - Utf8 version StringBuilder
- [Runtime T4 Generator](https://github.com/pcysl5edgo/runtimet4generator)
  - This generates code from .tt file
- [StringLiteralGenerator](https://github.com/ufcpp/StringLiteralGenerator)
  - This embeds Utf8 ReadOnlySpan&lt;byte&gt; literal

# Reasons of the each file change

## Deleteion of *Template.cs

- ResolverTemplate.cs
- EnumTemplate.cs
- UnionTemplate.cs
- FormatterTemplate.cs
- StringKeyFormatterTemplate.cs

They are not necessary anymore.

## MessagePack.Generator.csproj

`ConsoleAppFramework` version update.

## IFormatterTemplate.cs

Add new interface `ITemplate`.

## TemplatePartials.cs

`UnionTemplate`, `EnumTemplate` and `ResolverTemplate` implements `ITemplate`.
All Template partial classes' constructor have new parameter `CancellationToken cancellationToken`.

## MessagePack.GeneratorCore.csproj

### TargetFrameworks net6.0

[Visual Studio team plans to use net6.0 for their Roslyn process.](https://github.com/dotnet/roslyn/issues/45534)
I hope they will succeed this.

Anyway, `MessagePack.Generator` runs on net6.0. Adding net6.0 to `TargetFrameworks` is reasonable.

### AdditionalFiles

RuntimeT4Generator requires `AdditionalFiles`.

## CodeGenerator.cs

A lot of optimizations happened.

### Loop improvement

TypeCollector's Collect method has nothing to do with `multipleOutputSymbols`.
Put them out of the loop.

### 4 code paths

There are 4 paths in `GenerateFileAsync.GenerateFileAsync`.

||single file output|directory files output|
|-|-|-|
|no symbol|||
|multiple symbols|||

Every path is now optimized.

## CodeGeneratorTemplate.tt

Since 2 of the 4 paths(multiple symbols) are very similar, they are written in a .tt file.

## StringKeyFormatterDeserializeHelper.cs

Current code targets `StringBuilder`.
This PR code targets `Utf8ValueStringBuilder`.

neuecc commented:
I'm against it because it adds too many unnecessary dependencies.
Adding unnecessary dependencies in unimportant situations will increase the cost of future maintenance.

pCYSl5EDgo commented:
- Added Dependency
  - ZString
  - RuntimeT4Generator(Dev dependency)
  - StringLiteralGenerator(Dev dependency)
  - T4 normal generation(Dev dependency)
- Removed Dependency
  - T4 runtime generation(Dev dependency)
  - CodeDom

The balance is 2 dependencies to the bad.
Is it ok for you that the balance becomes 1 dependenct to good?

pCYSl5EDgo commented:
Dependencies are removed.

neuecc commented:
still exists RuntimeT4Generator.
and what is simplify? https://github.com/neuecc/MessagePack-CSharp/pull/1374/commits/65626bcccb2072eb1df1ca545d354cc50a822dc1
It is terrible to embed so many logics in a template.
It is hard to read and hard to debug.

pCYSl5EDgo commented:
Original code(I had written) is also hard to read...
Okay, I will remove the RuntimeT4Template dependency.

pCYSl5EDgo commented:
RuntimeT4Generator is now removed.

neuecc commented:
You are adding `MessagePack.Generator.T4`, it is dependency.

Does this have a significant benefit?
If not, then you are introducing an extra complication to the even more complicated templating.

Reducing memory allocation during code generations is not important I think.
Run-time optimization is important, but compile-time optimization is not.
It is more important to keep the code base clean.
I think the cleanliness of bringing in your own framework, which only you understand, will worsen maintainability.

pCYSl5EDgo commented:
Yes.
Indent T4 template is siginificantly useful.

Current T4 template cannot be embedded in nested indented places easily.

The coder must handle indent and StringBuilder manually.
It makes the code (such as StringFormatterDeserializeHelper.cs) unreadable.

`MessagePack.Generator.T4` and its `Indent T4 Template` (DeserializeHelperRecursiveTemplate.tt) is now just a T4 template.

AArnott commented:
I haven't reviewed the code changes, but given the PR discussion on dependencies, just a couple thoughts:
1. What issues do additional dependencies in mpc cause? mpc is a tool we ship separately, so I wouldn't expect dependencies to be a problem for users anyway.
2. Is it really worth investing further in T4 template generation when we probably want to be moving in the Source Generator direction going forward? Of course, we *could* use T4 templates as the basis for the source code that Source Generators introduce into the compilation, but I would wonder if rewriting in ordinary C# or using syntax trees directly would be simpler.

pCYSl5EDgo commented:
1. Dependencies Problem
`MessagePack.GeneratorCore` is not publicly Nuget-packaged.
The end users are not affected by the development time dependencies.

2. The value of T4 runtime template generation
The current Visual Studio's T4 has a limitation in indentation.
It is better to rewrite in ordinal C# than to use the current VS's T4.

AArnott commented:
A tool like mpc shouldn't have nuget dependencies anyway. Tools are expected to carry all their dependencies within their own package.
